### PR TITLE
CI-13300 Build Intelligence docs

### PR DESCRIPTION
**Pull request by dewan-ahmed**
This PR adds the build intelligence docs for Harness CI. Since this is a beta feature, this PR adds a `Harness CI Beta Feature` pill to the page.

1. Gradle and Bazel's cache service endpoint will vary if the customer is not on prod1/prod2. How do we inform them of where to find those endpoints? Feel free to commit to the doc directly.
2. For Bazel, I'm waiting for the proxy binary link from @rutvijmehta-harness. Once I have it, I can add it to the docs.
3. For `HARNESS_CACHE_SERVER_API_TOKEN`, I believe this will be a SAT instead of a PAT. Please confirm.
4. For "account admin (or edit) permission", I couldn't find a docs page that lists this specific permission. Could you please point me to a page that I can refer here?
5. I didn't include the entire script for Bazel config based on my discussion with @rutvijmehta-harness. Please check build-intelligence#how-it-works-1 and provide feedback on what needs to be updated.
6. Same comment on the final point $ENDPOINT:$PORT . We need a way to tell the customer where they can find these endpoints for their clusters. 

In addition to the above, please suggest any other fixes or places on HDH where we need to add content for Build Intelligence.

Checklist:

- [x] Successful preview build.
- [x] Code owner review.
- [x] No merge conflicts.
- [ ] Release notes/new features docs: Feature/version released to at least one prod environment.

### CODE REVIEW

File: docs/continuous-integration/get-started/harness-ci-intelligence.md [hunk 1]
```diff
@@ -61,8 +61,16 @@ You can use DLC with any [build infrastructure](/docs/continuous-integration/use
...
+|-----------------------|--------------------------------------------------------------------------------------------------------------|-------------------------------------------------|------------------------------|-------------------------------------------------------------------------------------------------------|
+| **Docker Layer Caching (DLC)**   | Reuses previously built Docker image layers, building only what’s necessary.                               | Speeds up Docker image build and push steps     | Docker images                | Can be used with any build infrastructure; cache stored in Harness-managed environment or S3-compatible storage. |
+| **Cache Intelligence**     | Caches project dependencies (e.g., plugins) downloaded at the start of the build in a specific directory. | Caches dependencies to save download time       | Project dependencies         | Automatically caches and restores dependencies; works with any build infrastructure; stored in Harness-managed or S3-compatible storage. |
```
Comment (neo413001) [inline docs/continuous-integration/get-started/harness-ci-intelligence.md:75]: Software dependencies instead of Project dependencies?

File: docs/continuous-integration/get-started/harness-ci-intelligence.md [hunk 2]
```diff
@@ -61,8 +61,16 @@ You can use DLC with any [build infrastructure](/docs/continuous-integration/use
...
+| **Build Intelligence**            | Caches outputs of cacheable build tasks as identified by the build tool (e.g., Gradle tasks).            | Speeds up overall build process by reusing task outputs | Build outputs                | Available for Gradle and Bazel, with Maven support coming soon; stores outputs locally or remotely.      |
```
Comment (neo413001) [inline docs/continuous-integration/get-started/harness-ci-intelligence.md:76]: Do we want to mention that this is supported in Harness Cloud only with support for self-hosted coming soon?

File: docs/continuous-integration/use-ci/build-and-upload-artifacts/build-intelligence.md [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+import Tabs from '@theme/Tabs';
+import TabItem from '@theme/TabItem';
+
+<DocsTag backgroundColor= "#cbf542" text="Harness CI Beta Feature" textColor="#4e42f5" />
+
+Build Intelligence is part of the suite of intelligent features in Harness CI designed to improve build times. It saves time by reusing outputs from previous builds. BI works by storing these outputs locally or remotely and retrieving them when inputs haven't changed. This process avoids the need to regenerate outputs, significantly speeding up the build process and enhancing efficiency.
+
+Build Intelligence in Harness CI is currently available for **Gradle** and **Bazel** with **Maven** support coming soon.
+
```
Comment (neo413001) [inline docs/continuous-integration/use-ci/build-and-upload-artifacts/build-intelligence.md:15]: It would be good to mention all the languages supported as well.

**Comment by rutvijmehta-harness** [comment]
### Proxy link
`https://app.harness.io/storage/harness-download/harness-ti/cache-proxy/<os>/<arch>/cache-proxy`

Options:
1. linux/amd64
2. linux/arm
3. mac/amd64
4. mac/arm
5. windows/amd64

Example:
https://app.harness.io/storage/harness-download/harness-ti/cache-proxy/linux/amd64/cache-proxy

### Answers to the above set of questions
1. Let's see if there's precedence for this in our docs
2. Proxy link mentioned above
3. Both SAT and PAT work
4. `core_account_edit` is the permission we look for
5. Currently reviewing it
6. Same as answer 1

File: docs/continuous-integration/use-ci/build-and-upload-artifacts/build-intelligence.md (hunk 3 above)
Comment (dewan-ahmed) [reply docs/continuous-integration/use-ci/build-and-upload-artifacts/build-intelligence.md:13]: Gradle primarily supports the following languages:

- Java
- Groovy
- Kotlin
- Scala
- C++
- C
- Objective-C
- Objective-C++
- Swift

Gradle also supports other languages and frameworks through plugins:

- Android
- JavaScript
- TypeScript
- Python (via third-party plugins)
- Ruby (via third-party plugins)

Bazel

- Java
- C++
- Python
- Go
- JavaScript
- TypeScript
- Objective-C
- Swift
- Rust
- Dart (through Flutter)
---
it might be a long list to mention on the docs. do you see a summarized version of this list we can mention?

File: docs/continuous-integration/use-ci/build-and-upload-artifacts/build-intelligence.md [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+Since Bazel does not support writing a custom plugin, the cache proxy exposes the contract. It will talk to cache service to get the pre-signed URLs and upload/download will be done in the cache proxy using those pre-signed URLs.
+2. Start the proxy server before the bazel command. You need to set the following environment variables:
+
+```bash
+HARNESS_CACHE_PROXY_ENABLED=true
+HARNESS_CACHE_SERVER_URL=
+HARNESS_CACHE_SERVER_API_TOKEN=
+```
+
+For `HARNESS_CACHE_SERVER_URL`, you can get the URL from FF or gateway endpoint if using prod1 or prod2 clusters.
```
Comment (hemanthmantri) [inline docs/continuous-integration/use-ci/build-and-upload-artifacts/build-intelligence.md:152]: It is not clear.

File: docs/continuous-integration/use-ci/build-and-upload-artifacts/build-intelligence.md [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+then
+    # Start the proxy server
+    mkdir -p /tmp/cachebin
+    curl -L -o /tmp/cachebin/harness-cache $HARNESS_CACHE_SERVER_URL
+    chmod +x /tmp/cachebin/harness-cache
+    echo "Starting cache proxy with account $HARNESS_ACCOUNT_ID"
+    /tmp/cachebin/harness-cache server > /tmp/cachebin/server.log 2>&1 &
+```
+
+3. Add the build cache endpoint to the relevant `bazelrc` file:
```
Comment (hemanthmantri) [inline docs/continuous-integration/use-ci/build-and-upload-artifacts/build-intelligence.md:169]: @rutvijmehta-harness We should also have a way to run it withoyt bazelrc file, right? Users can just use a build command without rc files.

File: docs/continuous-integration/use-ci/build-and-upload-artifacts/build-intelligence.md [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+                  name: SAMPLE_Test
+                  spec:
+                    connectorRef: <+input>
+                    image: <+input>
+                    shell: Sh
+                    command: gradle --build-cache unitTest -PmaxParallelForks=16 -PignoreFailures=true --profile
+                    envVariables:
+                      HARNESS_PAT: <+secrets.getValue("HARNESS_API_TOKEN")>
+                      CI: "true"
+                    intelligenceMode: true
```
Comment (hemanthmantri) [inline docs/continuous-integration/use-ci/build-and-upload-artifacts/build-intelligence.md:112]: This is not related to Build intelligence.

**Review by hemanthmantri — DISMISSED** [review]

File: docs/continuous-integration/use-ci/build-and-upload-artifacts/build-intelligence.md (hunk 4 above)
Comment (dewan-ahmed) [reply docs/continuous-integration/use-ci/build-and-upload-artifacts/build-intelligence.md:152]: Agreed. Please suggest where the customer can get the URL from. This is one of the questions I raised in the PR description.

File: docs/continuous-integration/use-ci/build-and-upload-artifacts/build-intelligence.md (hunk 6 above)
Comment (dewan-ahmed) [reply docs/continuous-integration/use-ci/build-and-upload-artifacts/build-intelligence.md:112]: Ack. Will remove these 4 lines.

File: docs/continuous-integration/use-ci/build-and-upload-artifacts/build-intelligence.md [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+
+1. Proxy binary: Download the proxy binary from `https://app.harness.io/storage/harness-download/harness-ti/cache-proxy/OS/ARCH/cache-proxy`. Replace **OS/ARCH** in the URL with one of the following options:
+
+- linux/amd64
+- linux/arm
+- mac/amd64
+- mac/arm
+- windows/amd64
+
+Since Bazel does not support writing a custom plugin, the cache proxy exposes the contract. It will talk to cache service to get the pre-signed URLs and upload/download will be done in the cache proxy using those pre-signed URLs.
```
Comment (rutvijmehta-harness) [inline docs/continuous-integration/use-ci/build-and-upload-artifacts/build-intelligence.md:143]: Do we need to give this implementation detail? We can be vague about the proxy server securely uploading and downloading cache blobs.

File: docs/continuous-integration/use-ci/build-and-upload-artifacts/build-intelligence.md [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+For `HARNESS_CACHE_SERVER_API_TOKEN`, create a Harness [Personal Access Token or Service Account Token](https://developer.harness.io/docs/platform/automation/api/add-and-manage-api-keys) with `core_account_edit` permission.
+
+Here is a sample script:
+
+```bash
+if [[ ! -z "${ENABLE_CI_CACHE// }" ]] && [[ "$ENABLE_CI_CACHE" == "true" ]]
+then
+    # Start the proxy server
+    mkdir -p /tmp/cachebin
+    curl -L -o /tmp/cachebin/harness-cache $HARNESS_CACHE_SERVER_URL
```
Comment (rutvijmehta-harness) [inline docs/continuous-integration/use-ci/build-and-upload-artifacts/build-intelligence.md:163]:
```suggestion
    curl -L -o /tmp/cachebin/harness-cache <the proxy download URL mentioned above>
```

File: docs/continuous-integration/use-ci/build-and-upload-artifacts/build-intelligence.md (hunk 5 above)
Comment (rutvijmehta-harness) [reply docs/continuous-integration/use-ci/build-and-upload-artifacts/build-intelligence.md:169]: Yes. Here's a sample bazel command:
`bazel build --remote_cache=http://localhost:8082/cache/bazel //...`

File: docs/continuous-integration/use-ci/build-and-upload-artifacts/build-intelligence.md [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+    chmod +x /tmp/cachebin/harness-cache
+    echo "Starting cache proxy with account $HARNESS_ACCOUNT_ID"
+    /tmp/cachebin/harness-cache server > /tmp/cachebin/server.log 2>&1 &
+```
+
+3. Add the build cache endpoint to the relevant `bazelrc` file: 
+
+```bash
+build --remote_cache=http://$HARNESS_CACHE_SERVER_URL:$PORT
```
Comment (rutvijmehta-harness) [inline docs/continuous-integration/use-ci/build-and-upload-artifacts/build-intelligence.md:172]:
```suggestion
build --remote_cache=http://localhost:8082/cache/bazel
```

File: docs/continuous-integration/use-ci/build-and-upload-artifacts/build-intelligence.md [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+
+```bash
+if [[ ! -z "${ENABLE_CI_CACHE// }" ]] && [[ "$ENABLE_CI_CACHE" == "true" ]]
+then
+    # Start the proxy server
+    mkdir -p /tmp/cachebin
+    curl -L -o /tmp/cachebin/harness-cache $HARNESS_CACHE_SERVER_URL
+    chmod +x /tmp/cachebin/harness-cache
+    echo "Starting cache proxy with account $HARNESS_ACCOUNT_ID"
+    /tmp/cachebin/harness-cache server > /tmp/cachebin/server.log 2>&1 &
```
Comment (rutvijmehta-harness) [inline docs/continuous-integration/use-ci/build-and-upload-artifacts/build-intelligence.md:166]: We need to end the `if` condition here using `fi` IIRC

File: docs/continuous-integration/use-ci/build-and-upload-artifacts/build-intelligence.md [hunk 11]
```diff
@@ -0,0 +1,176 @@
+---
+title: Build Intelligence Overview
+description: Learn about the build intelligence feature in Harness CI.
+sidebar_position: 7
+---
+
+import Tabs from '@theme/Tabs';
+import TabItem from '@theme/TabItem';
+
+Build Intelligence is part of the suite of intelligent features in Harness CI designed to improve build times. It saves time by reusing outputs from previous builds. BI works by storing these outputs locally or remotely and retrieving them when inputs haven't changed. This process avoids the need to regenerate outputs, significantly speeding up the build process and enhancing efficiency.
```
Comment (nofarb) [inline docs/continuous-integration/use-ci/build-and-upload-artifacts/build-intelligence.md:10]: change 
Build Intelligence is part of the suite of intelligent features in Harness CI designed to improve build times.
to
Build Intelligence is part of Harness CI Intelligence, a suite of features designed to improve build times.
please make  'Harness CI Intelligence' a link to the intelligence page

File: docs/continuous-integration/use-ci/build-and-upload-artifacts/build-intelligence.md [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+
+Build Intelligence in Harness CI is currently available for **Gradle** and **Bazel** with **Maven** support coming soon.
```
Comment (nofarb) [inline docs/continuous-integration/use-ci/build-and-upload-artifacts/build-intelligence.md:12]: change 
Build Intelligence in Harness CI is currently available for Gradle and Bazel with Maven support coming soon.

to 
Build Intelligence in Harness CI is currently available for Gradle and Bazel with Maven support coming soon. Regardless of the programing language used in your projects, as long as you're building with a supported build tool, you can take advantage of Build Intelligence to optimize your builds.

File: docs/continuous-integration/use-ci/build-and-upload-artifacts/build-intelligence.md (hunk 11 above)
Comment (nofarb) [inline docs/continuous-integration/use-ci/build-and-upload-artifacts/build-intelligence.md:10]: please add a note at the top that this is behind a FF , contact support to enable, and make it clear that it's currently only offered to cloud, with support for self hosted coming soon

Comment (nofarb) [reply docs/continuous-integration/use-ci/build-and-upload-artifacts/build-intelligence.md:10]: similarly to the not we have on dlc page

File: docs/continuous-integration/use-ci/build-and-upload-artifacts/build-intelligence.md (hunk 7 above)
Comment (dewan-ahmed) [reply docs/continuous-integration/use-ci/build-and-upload-artifacts/build-intelligence.md:143]: Replaced with:

The cache proxy facilitates secure uploading and downloading of cache blobs by interacting with the cache service to obtain necessary access URLs.

File: docs/continuous-integration/use-ci/build-and-upload-artifacts/build-intelligence.md (hunk 8 above)
Comment (dewan-ahmed) [reply docs/continuous-integration/use-ci/build-and-upload-artifacts/build-intelligence.md:163]: Hi @rutvijmehta-harness which proxy download URL? Could you pls point me the line # in this doc?
> the proxy download URL mentioned above

File: docs/continuous-integration/use-ci/build-and-upload-artifacts/build-intelligence.md (hunk 11 above)
Comment (dewan-ahmed) [reply docs/continuous-integration/use-ci/build-and-upload-artifacts/build-intelligence.md:10]: What's the FF for build intelligence?

**Comment by dewan-ahmed** [comment]
Hi @nofarb @rutvijmehta-harness I addressed previous review comments via https://github.com/harness/developer-hub/pull/7537/commits/2925e9bf9d1f42be4a81af6ff848caf5f305d717

There are three open questions:
1. https://github.com/harness/developer-hub/pull/7537/#discussion_r1706148115
2. https://github.com/harness/developer-hub/pull/7537/#discussion_r1710251046
3. https://github.com/harness/developer-hub/pull/7537/#discussion_r1710279958

Once I get answers to those, I'll add the final changes. Thanks.

File: docs/continuous-integration/use-ci/build-and-upload-artifacts/build-intelligence.md (hunk 8 above)
Comment (rutvijmehta-harness) [reply docs/continuous-integration/use-ci/build-and-upload-artifacts/build-intelligence.md:166]: `https://app.harness.io/storage/harness-download/harness-ti/cache-proxy/OS/ARCH/cache-proxy`

File: docs/continuous-integration/use-ci/build-and-upload-artifacts/build-intelligence.md (hunk 4 above)
Comment (dewan-ahmed) [reply docs/continuous-integration/use-ci/build-and-upload-artifacts/build-intelligence.md]: HARNESS_CACHE_SERVER_URL is https://app.harness.io/gateway/cache-service for prod1 or prod2 clusters

**Comment by dewan-ahmed** [comment]
based on today's CI standup, this PR is good to get merged.

**Review by lkilpatrick-harness — APPROVED** [review]
LGTM
